Custom-ESLint-checks.md

```diff
@@ -0,0 +1,329 @@
+
+#### Write the rule
+
+To write the rule, enable "Transform" on AST Explorer.
+
+ESLint rules are defined as functions that return objects. The returned objects have keys for different types of nodes, and each key maps to a function that checks whether a node of that type violates the rule. In our case, we want to check whether nodes of type `Literal` have a `value` attribute equal to `Hello, world!`, so we can write this rule in AST Explorer's bottom-left panel:
+
+```js
+export default function(context) {
+  return {
+    Literal: function(node) {
+      if (node.value === "Hello, world!") {
+        context.report({
+          node,
+          message: 'Do not use "Hello, world!" strings',
+        });
+      }
+    }
+  };
+};
+```
+
+Notice that the node we are examining gets passed into our function and that we report errors by calling `context.report`. In the lower right on AST Explorer, you can see that this lint rule works as expected:
+
+![Screenshot of AST Explorer with the lint rule running on code with two "Hello, world!" literals and one constructed string. The linter output flags only the literals.](images/customEslintChecks/helloWorldRule.png)
+
+#### Add the rule to Oppia
+
+To add the rule to Oppia's suite of lint checks, we need to add some more metadata describing each problem the rule can find. For our hello world rule, we would create a JavaScript file in `scripts/linters/custom_eslint_checks/rules/hello-world.js` with the following code:
+
+```js
+'use strict';
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: '"Hello, world!" strings should not be used',
+      category: 'Best Practices',
+      recommended: true,
+    },
+    fixable: null,
+    schema: [],
+    messages: {
+      helloWorld: 'Do not use "Hello, world!" strings',
+    },
+  },
+
+  create: function(context) {
+    return {
+      Literal: function(node) {
+        if (node.value === "Hello, world!") {
+          context.report({
+            node: node,
+            messageId: 'helloWorld',
+          });
+        }
+      }
+    };
+  }
+};
+```
+
+First, consider the `meta` key at the top of the exported object. The associated object contains the rule's metadata. Under `docs` we describe the rule, and under `messages` we list the error messages that this rule can raise. ESLint supports rules that suggest fixes, but since we can't automatically fix a `Hello, world!` usage, we set `fixable` to `null`.
+
+Next, consider the `create` key. This key maps to the function we wrote in AST Explorer. However, notice that the values passed to `context.report()` have changed. Instead of specifying the error message directly, it's better to reference a message from the rule's metadata.
```

Comment:
Some questions I have when reading this:

- Is the key always called create or is that a custom name that the author can choose?
- More generally, which parts of this structure are fixed and which are left to the discretion of the author? I.e. what is the actual "spec" for such a rule? (The discussion so far is about a specific example so I'm not sure what stuff falls in "general" vs "example" category.)
- It took me a while and some cross-comparison to puzzle out `However, notice that the values passed to `context.report()` have changed. Instead of specifying the error message directly, it's better to reference a message from the rule's metadata.`. Maybe explain specifically which key has changed and how to compute/specify the messageId?

Reply:
done